### DISCUSSION TITLE
Semantic concepts support?

### DISCUSSION THREAD

**Discussion by domino-pablo-rodriguez** [cue-lang/cue#1890] [Ideas]
I’m currently evaluating CUE to use it for both yaml validation and semantic inference (similar to https://a.ml/docs/). From the documentation it seems like assigning a semantic meaning and querying is [not supported](https://cuelang.org/docs/usecases/query/). An example would be: be able to mark certaing bits of a yaml/json configuration file as a “Secret” (and define other concepts that can inherit from this, such as a password, token, keys, etc) and be able to query them (for example to redact them). This semantic meaning is orthogonal to the schema values (a string could be a secret, or a more complex struct). So my question is (maybe to the owners of the spec), do you envision that CUE could fit this use case?

**Comment by eadlam** [comment]
@domino-pablo-rodriguez,  Are you asking if cue can consume json/yaml and produce new json/yaml with additional fields based on the original?

**main.cue**
```cue
package main

// stub for imported yaml/josn
let data = {
    name: "Guy"
    email: "guy@gmail.com"
    secret: "onetwothree"
}

// Extensions
{
    data
    extensions: {
        password: "pw_\(data.secret)"
    }
}
```
**cli**
```
$ cue export example.com:main
{
    "extensions": {
        "password": "pw_onetwothree"
    },
    "name": "Guy",
    "email": "guy@gmail.com",
    "secret": "onetwothree"
}
```

**Comment by myitcv** [comment]
@domino-pablo-rodriguez I think you are looking for [attributes](https://tip.cuelang.org/docs/references/spec/#attributes), but if you can expand with an example (in pseudo code) then we might be able to help further.
